core/src/main/java/bisq/core/xmr/connection/MoneroConnectionsManager.java

```diff
@@ -0,0 +1,175 @@
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+
+import monero.common.MoneroConnectionManager;
+import monero.common.MoneroRpcConnection;
+
+@Slf4j
+@Singleton
+public final class MoneroConnectionsManager {
+
+    private final Object lock = new Object();
+
+    private final MoneroConnectionManager connectionManager;
+    private final XmrConnectionList connectionList;
+
+    @Inject
+    public MoneroConnectionsManager(MoneroConnectionManager connectionManager,
+                                    XmrConnectionList connectionList) {
+        this.connectionManager = connectionManager;
+        this.connectionList = connectionList;
+        initialize();
+    }
+
+    private void initialize() {
+        synchronized (lock) {
+            loadConnections();
+            addDefaultConnection();
+        }
+    }
+
+    private void loadConnections() {
+        connectionList.getConnections().forEach(uriConnection ->
+                connectionManager.addConnection(toMoneroRpcConnection(uriConnection)));
+    }
+
+    private void addDefaultConnection() {
+        String defaultUri = WalletConfig.MONERO_DAEMON_URI;
+        if (!connectionList.hasConnection(defaultUri)) {
+            addConnection(UriConnection.builder()
+                    .uri(defaultUri)
+                    .username(WalletConfig.MONERO_DAEMON_USERNAME)
+                    .password(WalletConfig.MONERO_DAEMON_PASSWORD)
+                    .build());
+        }
+    }
+
+    public void addConnection(UriConnection connection) {
+        synchronized (lock) {
+            connectionList.addConnection(connection);
+            connectionManager.addConnection(toMoneroRpcConnection(connection));
+        }
+    }
+
+    public void removeConnection(String uri) {
+        synchronized (lock) {
+            connectionList.removeConnection(uri);
+            connectionManager.removeConnection(uri);
+        }
+    }
+
+    public UriConnection getConnection() {
+        synchronized (lock) {
+            return toUriConnection(connectionManager.getConnection());
+        }
+    }
+
+    public List<UriConnection> getConnections() {
+        synchronized (lock) {
+            return connectionManager.getConnections().stream().map(this::toUriConnection).collect(Collectors.toList());
+        }
+    }
+
+    public void setConnection(String connectionUri) {
+        synchronized (lock) {
+            connectionManager.setConnection(connectionUri);
+        }
+    }
+
+    public void setConnection(UriConnection connection) {
+        synchronized (lock) {
+            if (connectionManager.getConnectionByUri(connection.getUri()) != null
+                    && connection.getUsername() == null
+                    && connection.getPassword() == null) {
+                // If the connection by this uri is already known, and nothing of special value is set,
```

Comment:
Please have a look at this implementation. I don't think there is a better way, to switch to an existing connection with configured username/password without resetting them.

Reply:
Right, if credentials are not provided, it should select the previously added connection. If credentials are provided, it should overwrite the current credentials.

If we want to clear credentials over the API, currently the connection needs to be removed and re-added.

Thinking about this further, maybe we should actually keep the two grpc api calls, setConnection(uri) and setConnection(uri, username, password), to avoid needing to remove the connection in order to clear the credentials.